File: src/routes/projects.ts
import { Hono } from 'hono'
import { eq, desc, count } from 'drizzle-orm'
import { db } from '../db/index.js'
import { projects, lines } from '../db/schema.js'
import { serializeLine } from './serialize.js'

const app = new Hono()

// 一覧: リストペイン用に行数と updated_at を付ける
app.get('/api/projects', (c) => {
  const rows = db
    .select({
      id: projects.id,
      name: projects.name,
      created_at: projects.created_at,
      updated_at: projects.updated_at,
      lineCount: count(lines.id),
    })
    .from(projects)
    .leftJoin(lines, eq(lines.project_id, projects.id))
    .groupBy(projects.id)
    .orderBy(desc(projects.updated_at))
    .all()
  return c.json(rows)
})

// 作成
app.post('/api/projects', async (c) => {
  const body = await c.req.json().catch(() => ({}))
  const name = typeof body.name === 'string' ? body.name.trim() : ''
  if (!name) return c.json({ error: 'name required' }, 400)

  const row = db.insert(projects).values({ name }).returning().get()
  return c.json(row, 201)
})

// 単体 + その行を position 昇順で
app.get('/api/projects/:id', (c) => {
  const id = Number(c.req.param('id'))
  if (Number.isNaN(id)) return c.json({ error: 'invalid id' }, 400)

  const project = db.select().from(projects).where(eq(projects.id, id)).get()
  if (!project) return c.json({ error: 'not found' }, 404)

  const rows = db
    .select()
    .from(lines)
    .where(eq(lines.project_id, id))
    .orderBy(lines.position)
    .all()
  return c.json({ ...project, lines: rows.map(serializeLine) })
})

// 改名 (updated_at 更新)
app.patch('/api/projects/:id', async (c) => {
  const id = Number(c.req.param('id'))
  if (Number.isNaN(id)) return c.json({ error: 'invalid id' }, 400)

  const body = await c.req.json().catch(() => ({}))
  const name = typeof body.name === 'string' ? body.name.trim() : ''
  if (!name) return c.json({ error: 'name required' }, 400)

  const row = db
    .update(projects)
    .set({ name, updated_at: new Date().toISOString() })
    .where(eq(projects.id, id))
    .returning()
    .get()
  if (!row) return c.json({ error: 'not found' }, 404)
  return c.json(row)
})

// 削除 (lines も cascade 削除)
app.delete('/api/projects/:id', (c) => {
  const id = Number(c.req.param('id'))
  if (Number.isNaN(id)) return c.json({ error: 'invalid id' }, 400)

  const project = db.select().from(projects).where(eq(projects.id, id)).get()
  if (!project) return c.json({ error: 'not found' }, 404)

  db.transaction((tx) => {
    tx.delete(lines).where(eq(lines.project_id, id)).run()
    tx.delete(projects).where(eq(projects.id, id)).run()
  })
  return c.json({ ok: true })
})

export default app

File: src/routes/projects.rs
// プロジェクト CRUD。TS 版 src-ts/routes/projects.ts を契約互換に移植。

use crate::db::now_iso;
use crate::error::AppError;
use crate::serialize::{lines_by_project, serialize_line};
use crate::state::AppState;
use axum::body::Bytes;
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::routing::get;
use axum::{Json, Router};
use rusqlite::OptionalExtension;
use serde_json::{json, Value};

use super::{parse_body, parse_id, JsonResult};

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/api/projects", get(list).post(create))
        .route("/api/projects/{id}", get(get_one).patch(rename).delete(remove))
}

// project 行 (全列) を Value にする。
fn project_value(
    id: i64,
    name: String,
    created_at: String,
    updated_at: String,
) -> Value {
    json!({ "id": id, "name": name, "created_at": created_at, "updated_at": updated_at })
}

// 一覧: リストペイン用に行数と updated_at を付ける
async fn list(State(state): State<AppState>) -> JsonResult {
    let conn = state.db.lock().unwrap();
    let mut stmt = conn.prepare(
        "SELECT projects.id, projects.name, projects.created_at, projects.updated_at, \
         count(lines.id) AS lineCount \
         FROM projects LEFT JOIN lines ON lines.project_id = projects.id \
         GROUP BY projects.id ORDER BY projects.updated_at DESC",
    )?;
    let rows = stmt
        .query_map([], |r| {
            Ok(json!({
                "id": r.get::<_, i64>(0)?,
                "name": r.get::<_, String>(1)?,
                "created_at": r.get::<_, String>(2)?,
                "updated_at": r.get::<_, String>(3)?,
                "lineCount": r.get::<_, i64>(4)?,
            }))
        })?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    Ok((StatusCode::OK, Json(Value::Array(rows))))
}

// 作成
async fn create(State(state): State<AppState>, body: Bytes) -> JsonResult {
    let body = parse_body(&body);
    let name = body
        .get("name")
        .and_then(Value::as_str)
        .map(str::trim)
        .unwrap_or("");
    if name.is_empty() {
        return Err(AppError::BadRequest("name required".into()));
    }
    let now = now_iso();
    let row = {
        let conn = state.db.lock().unwrap();
        conn.query_row(
            "INSERT INTO projects (name, created_at, updated_at) VALUES (?1, ?2, ?3) \
             RETURNING id, name, created_at, updated_at",
            (name, &now, &now),
            |r| Ok(project_value(r.get(0)?, r.get(1)?, r.get(2)?, r.get(3)?)),
        )?
    };
    Ok((StatusCode::CREATED, Json(row)))
}

// 単体 + その行を position 昇順で
async fn get_one(State(state): State<AppState>, Path(id): Path<String>) -> JsonResult {
    let id = parse_id(&id)?;
    let conn = state.db.lock().unwrap();
    let project = conn
        .query_row(
            "SELECT id, name, created_at, updated_at FROM projects WHERE id = ?1",
            [id],
            |r| Ok(project_value(r.get(0)?, r.get(1)?, r.get(2)?, r.get(3)?)),
        )
        .optional()?;
    let Some(mut project) = project else {
        return Err(AppError::NotFound("not found".into()));
    };
    let lines: Vec<Value> = lines_by_project(&conn, id)?
        .iter()
        .map(serialize_line)
        .collect();
    if let Value::Object(map) = &mut project {
        map.insert("lines".into(), Value::Array(lines));
    }
    Ok((StatusCode::OK, Json(project)))
}

// 改名 (updated_at 更新)
async fn rename(
    State(state): State<AppState>,
    Path(id): Path<String>,
    body: Bytes,
) -> JsonResult {
    let id = parse_id(&id)?;
    let body = parse_body(&body);
    let name = body
        .get("name")
        .and_then(Value::as_str)
        .map(str::trim)
        .unwrap_or("");
    if name.is_empty() {
        return Err(AppError::BadRequest("name required".into()));
    }
    let now = now_iso();
    let row = {
        let conn = state.db.lock().unwrap();
        conn.query_row(
            "UPDATE projects SET name = ?1, updated_at = ?2 WHERE id = ?3 \
             RETURNING id, name, created_at, updated_at",
            (name, &now, id),
            |r| Ok(project_value(r.get(0)?, r.get(1)?, r.get(2)?, r.get(3)?)),
        )
        .optional()?
    };
    match row {
        Some(row) => Ok((StatusCode::OK, Json(row))),
        None => Err(AppError::NotFound("not found".into())),
    }
}

// 削除 (lines も cascade 削除)
async fn remove(State(state): State<AppState>, Path(id): Path<String>) -> JsonResult {
    let id = parse_id(&id)?;
    let mut conn = state.db.lock().unwrap();
    let exists: Option<i64> = conn
        .query_row("SELECT id FROM projects WHERE id = ?1", [id], |r| r.get(0))
        .optional()?;
    if exists.is_none() {
        return Err(AppError::NotFound("not found".into()));
    }
    let tx = conn.transaction()?;
    tx.execute("DELETE FROM lines WHERE project_id = ?1", [id])?;
    tx.execute("DELETE FROM projects WHERE id = ?1", [id])?;
    tx.commit()?;
    Ok((StatusCode::OK, Json(json!({ "ok": true }))))
}
